### PR TITLE
ISSUE-2067: reduce byte[] allocation in add entry

### PR DESCRIPTION
**Pull request by mtang01**
Descriptions of the changes in this PR:
This change removes a byte[] copy in DigestManager digest calculation
(computeDigestAndPackageForSending)
that puts crc header and payload in a continuous buffer. Instead,
it uses protobuf ByteString.concat to concatenate header and payload
without copy when building protobuf message.

### Motivation

In add entry code path, I see lots of byte[] allocated to do digest calculation.  It's possible to not allocate byte[]. 

### Changes

Don't allocate a ByteBuf to copy data. Keep header and data separate, but use ByteString.concat when construct protobuf message.

Master Issue: #2067 

> ---
> In order to uphold a high standard for quality for code contributions, Apache BookKeeper runs various precommit
> checks for pull requests. A pull request can only be merged when it passes precommit checks. However running all
> the precommit checks can take a long time, some trivial changes don't need to run all the precommit checks. You
> can check following list to skip the tests that don't need to run for your pull request. Leave them unchecked if
> you are not sure, committers will help you:
>
> - [ ] [skip bookkeeper-server bookie tests]: skip testing `org.apache.bookkeeper.bookie` in bookkeeper-server module.
> - [ ] [skip bookkeeper-server client tests]: skip testing `org.apache.bookkeeper.client` in bookkeeper-server module.
> - [ ] [skip bookkeeper-server replication tests]: skip testing `org.apache.bookkeeper.replication` in bookkeeper-server module.
> - [ ] [skip bookkeeper-server tls tests]: skip testing `org.apache.bookkeeper.tls` in bookkeeper-server module.
> - [ ] [skip bookkeeper-server remaining tests]: skip testing all other tests in bookkeeper-server module.
> - [ ] [skip integration tests]: skip docker based integration tests. if you make java code changes, you shouldn't skip integration tests.
> - [ ] [skip build java8]: skip build on java8. *ONLY* skip this when *ONLY* changing files under documentation under `site`.
> - [ ] [skip build java11]: skip build on java11. *ONLY* skip this when *ONLY* changing files under documentation under `site`.
> ---

> ---
> Be sure to do all of the following to help us incorporate your contribution
> quickly and easily:
>
> If this PR is a BookKeeper Proposal (BP):
>
> - [ ] Make sure the PR title is formatted like:
>     `<BP-#>: Description of bookkeeper proposal`
>     `e.g. BP-1: 64 bits ledger is support`
> - [ ] Attach the master issue link in the description of this PR.
> - [ ] Attach the google doc link if the BP is written in Google Doc.
>
> Otherwise:
> 
> - [ ] Make sure the PR title is formatted like:
>     `<Issue #>: Description of pull request`
>     `e.g. Issue 123: Description ...`
> - [ ] Make sure tests pass via `mvn clean apache-rat:check install spotbugs:check`.
> - [ ] Replace `<Issue #>` in the title with the actual Issue number.
> 
> ---

### CODE REVIEW

**Comment by sijie** [comment]
run bookkeeper-server remaining tests
run integration tests

**Comment by mtang01** [comment]
The failed jenkins build https://builds.apache.org/job/bookkeeper_precommit_remaining_tests/994/console is not related to this change. It's has been failing for a while. I'll try to fix the tests.

**Comment by sijie** [comment]
run bookkeeper-server remaining tests

**Comment by mtang01** [comment]
run bookkeeper-server remaining tests

**Comment by mtang01** [comment]
@sijie All tests passed. Can this PR be merged?

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/meta/TestLedgerMetadataSerDe.java [hunk 1]
```diff
@@ -130,6 +130,7 @@ public void testPeggedToV2SerDe() throws Exception {
             .newEnsembleEntry(0L, Lists.newArrayList(new BookieSocketAddress("192.0.2.1", 3181),
                                                      new BookieSocketAddress("192.0.2.2", 3181),
                                                      new BookieSocketAddress("192.0.2.3", 3181)))
+            .withMetadataFormatVersion(LedgerMetadataSerDe.METADATA_FORMAT_VERSION_2)
```
Comment (eolivelli) [inline bookkeeper-server/src/test/java/org/apache/bookkeeper/meta/TestLedgerMetadataSerDe.java]: This part seems related

Cc @sijie

Comment (mtang01) [reply bookkeeper-server/src/test/java/org/apache/bookkeeper/meta/TestLedgerMetadataSerDe.java]: You're correct that it's unrelated. Try to fix a test regression introduced by https://github.com/apache/bookkeeper/commit/407cb35e594e2334ae687f7c884651c7642349fa. It has been fixed separately. I'll revert this test change.

**Comment by mtang01** [comment]
run integration tests

**Comment by mtang01** [comment]
The test fix is already in master as part of https://github.com/apache/bookkeeper/commit/86bce12b58f997430f87771abb95854ee41a7248

**Comment by mtang01** [comment]
run integration tests

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/checksum/DigestManager.java [hunk 2]
```diff
@@ -99,41 +98,22 @@ public static DigestManager instantiate(long ledgerId, byte[] passwd, DigestType
...
-
-            update(sendBuffer);
-            update(data);
-            populateValueAndReset(sendBuffer);
-
-            sendBuffer.writeBytes(data, data.readerIndex(), data.readableBytes());
-            ReferenceCountUtil.release(data);
-
-            return ByteBufList.get(sendBuffer);
+            headersBuffer = Unpooled.buffer(METADATA_LENGTH + macCodeLength);
```
Comment (karanmehta93) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/checksum/DigestManager.java]: Should this be
`ByteBuf sendBuffer = Unpooled.buffer(METADATA_LENGTH + macCodeLength + data.readableBytes());` instead?

Comment (mtang01) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/checksum/DigestManager.java]: no, whole point of this change is not to allocate new byte[] to concatenate header and data

**Comment by mtang01** [comment]
can this be merged now that all test runs are clean?

**Comment by sijie** [comment]
@mtang01 Great contribution! Thank you so much.

**Pull request merged**